### Задание 1. 
Найдите посредством NumPy SVD для матрицы:

$$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$$

In [18]:
import numpy as np
np.set_printoptions(precision=2, suppress=True)

In [19]:
A = np.array([[1, 2, 0],
              [0, 0, 5],
              [3, -4, 2],
              [1, 6, 5],
              [0, 1, 0]])
print(f'Матрица A:\n{A}')

Матрица A:
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]


In [20]:
U, s, W = np.linalg.svd(A)

# Транспонируем матрицу W
V = W.T

# s - список диагональных элементов, его нужно привести к виду диагональной матрицы для наглядности
D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

In [21]:
print(f'Матрица D:\n{D}')

Матрица D:
[[8.82 0.   0.  ]
 [0.   6.14 0.  ]
 [0.   0.   2.53]
 [0.   0.   0.  ]
 [0.   0.   0.  ]]


In [22]:
print(f'Матрица U:\n{U}')

Матрица U:
[[ 0.17  0.16 -0.53 -0.8  -0.16]
 [ 0.39 -0.53  0.61 -0.43  0.03]
 [-0.14 -0.82 -0.52  0.14  0.07]
 [ 0.89  0.06 -0.25  0.38 -0.06]
 [ 0.08  0.11 -0.08 -0.11  0.98]]


In [23]:
# Убедимся, что она действительно ортогональна
print(np.dot(U.T, U))

[[ 1.  0. -0. -0. -0.]
 [ 0.  1.  0. -0.  0.]
 [-0.  0.  1. -0. -0.]
 [-0. -0. -0.  1. -0.]
 [-0.  0. -0. -0.  1.]]


In [24]:
print(f'Матрица V:\n{V}')

Матрица V:
[[ 0.07 -0.37 -0.93]
 [ 0.72  0.67 -0.21]
 [ 0.69 -0.65  0.31]]


In [25]:
# Убедимся, что она действительно ортогональна
print(np.dot(V.T, V))

[[ 1.  0.  0.]
 [ 0.  1. -0.]
 [ 0. -0.  1.]]


In [26]:
# Проведем проверку
print(np.dot(np.dot(U, D), V.T))

[[ 1.  2.  0.]
 [ 0. -0.  5.]
 [ 3. -4.  2.]
 [ 1.  6.  5.]
 [-0.  1.  0.]]


### Задание 2. 
Для матрицы из предыдущего задания найдите:

    а) евклидову норму;

Евклидова норма матрицы равна евклидовой норме диагональной матрицы из её сингулярных чисел $D$. Максимальное значение полученного отношения будет равно максимальному сингулярному числу $\mu_{max}$, и, принимая во внимание факт сортировки по убыванию сингулярных чисел, получим:

$$\left \| A \right \|_{E}=\mu_{1}.$$

In [27]:
max(s)

8.824868854820442

Ответ: $$\left \| A \right \|_{E}=8.82.$$

### Задание 2. 
Для матрицы из предыдущего задания найдите:
    
    б) норму Фробениуса.

Когда известно сингулярное разложение матрицы, её норма Фробениуса вычисляется как:

$$\left \| A \right \|_{F}=\sqrt{\sum_{k=1}^{r}\mu_{k}^{2}}.$$

In [28]:
np.linalg.norm(s)

11.045361017187261

In [29]:
np.linalg.norm(A, ord = 'fro')

11.045361017187261

Ответ: $$\left \| A \right \|_{F}=11.04.$$